sdk/src/main/java/com/google/cloud/dataflow/sdk/runners/inprocess/TestStreamEvaluatorFactory.java

```diff
@@ -0,0 +1,195 @@
+ * the License.
+ */
+
+package com.google.cloud.dataflow.sdk.runners.inprocess;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.cloud.dataflow.sdk.Pipeline;
+import com.google.cloud.dataflow.sdk.runners.PipelineRunner;
+import com.google.cloud.dataflow.sdk.runners.inprocess.InProcessPipelineRunner.CommittedBundle;
+import com.google.cloud.dataflow.sdk.runners.inprocess.InProcessPipelineRunner.UncommittedBundle;
+import com.google.cloud.dataflow.sdk.testing.TestStream;
+import com.google.cloud.dataflow.sdk.testing.TestStream.ElementEvent;
+import com.google.cloud.dataflow.sdk.testing.TestStream.Event;
+import com.google.cloud.dataflow.sdk.testing.TestStream.EventType;
+import com.google.cloud.dataflow.sdk.testing.TestStream.ProcessingTimeEvent;
+import com.google.cloud.dataflow.sdk.testing.TestStream.WatermarkEvent;
+import com.google.cloud.dataflow.sdk.transforms.AppliedPTransform;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.windowing.BoundedWindow;
+import com.google.cloud.dataflow.sdk.util.WindowedValue;
+import com.google.cloud.dataflow.sdk.util.WindowingStrategy;
+import com.google.cloud.dataflow.sdk.values.PBegin;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.dataflow.sdk.values.PCollection.IsBounded;
+import com.google.cloud.dataflow.sdk.values.PInput;
+import com.google.cloud.dataflow.sdk.values.POutput;
+import com.google.cloud.dataflow.sdk.values.TimestampedValue;
+import com.google.common.base.Supplier;
+import org.joda.time.Duration;
+import org.joda.time.Instant;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import javax.annotation.Nullable;
+
+/**
+ * The {@link TransformEvaluatorFactory} for the {@link TestStream} primitive.
+ */
+class TestStreamEvaluatorFactory implements TransformEvaluatorFactory {
+  private final AtomicBoolean inUse = new AtomicBoolean(false);
+  private final AtomicReference<Evaluator<?>> evaluator = new AtomicReference<>();
+
+  @Nullable
+  @Override
+  public <InputT> TransformEvaluator<InputT> forApplication(
+      AppliedPTransform<?, ?, ?> application,
+      @Nullable CommittedBundle<?> inputBundle,
+      InProcessEvaluationContext evaluationContext) throws Exception {
+    return createEvaluator((AppliedPTransform) application, evaluationContext);
+  }
+
+  private <InputT, OutputT> TransformEvaluator<? super InputT> createEvaluator(
+      AppliedPTransform<PBegin, PCollection<OutputT>, TestStream<OutputT>> application,
+      InProcessEvaluationContext evaluationContext) {
+    if (evaluator.get() == null) {
+      Evaluator<OutputT> createdEvaluator = new Evaluator<>(application, evaluationContext, inUse);
+      evaluator.compareAndSet(null, createdEvaluator);
+    }
+    if (inUse.compareAndSet(false, true)) {
```

Comment:
Race condition averted. This should be keyed by application.

Evaluators are stateful (so we can produce the correct outputs) so they have to be cached, but can't be split or parallelized otherwise we get output faster than is expected.